exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,528 @@
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common/crypto"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wshandler"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wsorderbook"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	ftxWSURL          = "wss://ftx.com/ws/"
+	ftxWebsocketTimer = 13 * time.Second
+	wsTicker          = "ticker"
+	wsTrades          = "trades"
+	wsOrderbook       = "orderbook"
+	wsMarkets         = "markets"
+	wsFills           = "fills"
+	wsOrders          = "orders"
+	wsUpdate          = "update"
+	wsPartial         = "partial"
+	subscribe         = "subscribe"
+	unsubscribe       = "unsubscribe"
+	blockPair         = "DOGE-PERP"
+)
+
+var obSuccess = make(map[currency.Pair]bool)
+
+// WsConnect connects to a websocket feed
+func (f *FTX) WsConnect() error {
+	if !f.Websocket.IsEnabled() || !f.IsEnabled() {
+		return errors.New(wshandler.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	err := f.WebsocketConn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	f.WebsocketConn.SetupPingHandler(wshandler.WebsocketPingHandler{
+		MessageType: websocket.PingMessage,
+		Delay:       ftxWebsocketTimer,
+	})
+	if f.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Connected to Websocket.\n", f.Name)
+	}
+	f.GenerateDefaultSubscriptions()
+	go f.wsReadData()
+	if f.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		err := f.WsAuth()
+		if err != nil {
+			f.Websocket.DataHandler <- err
+			f.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+		f.GenerateAuthSubscriptions()
+	}
+	return nil
+}
+
+// WsAuth sends an authentication message to receive auth data
+func (f *FTX) WsAuth() error {
+	intNonce := time.Now().UnixNano() / 1000000
+	strNonce := strconv.FormatInt(intNonce, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA256,
+		[]byte(strNonce+"websocket_login"),
+		[]byte(f.API.Credentials.Secret),
+	)
+	sign := crypto.HexEncodeToString(hmac)
+	req := Authenticate{Operation: "login",
+		Args: AuthenticationData{
+			Key:  f.API.Credentials.Key,
+			Sign: sign,
+			Time: intNonce,
+		},
+	}
+	return f.WebsocketConn.SendJSONMessage(req)
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (f *FTX) Subscribe(channelToSubscribe wshandler.WebsocketChannelSubscription) error {
+	if channelToSubscribe.Currency.Equal(currency.NewPairFromString(blockPair)) {
+		log.Warnf(log.Global, "subscription not allowed for %v due to invalid checksum provided by the exchange", channelToSubscribe.Currency)
+		return nil
+	}
```

Comment:
```go
if channelToSubscribe.Currency.Equal(currency.NewPairFromString(blockPair)) && channelToSubscribe.Channel == "orderbook" {
		log.Warnf(log.Global, "subscription not allowed for %v due to invalid orderbook checksum provided by the exchange", channelToSubscribe.Currency)
		return nil
	}
```

Reply:
So we don't purge ticker and trade subs as they are not affected by the checksum and we can still fall over to REST fetching of orderbook.